### PR TITLE
argocd-autopilot: init at 0.2.30

### PR DESCRIPTION
**Pull request by sagikazarmark**
###### Motivation for this change

A new component in the ArgoCD ecosystem to run ArgoCD itself.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (jonringer) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:6]: is the commit really necessary?

For the nixpkgs-update machinery, it most likely wont update this, and it will get out-of-date

Comment (sagikazarmark) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix:6]: Commit hash is built into the binary and is reported by the version command. I can replace it with "unknown", similarly to build date if that's better.

Comment (sagikazarmark) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix:6]: @jonringer what do you think would be the better option.

Comment (bryanasdev000) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix:6]: +1 for removing it and leaving just version, I was using this the same way and was way too boring to update the PR from bot.

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:5]: Please update to the latest (https://github.com/argoproj-labs/argocd-autopilot/releases/tag/v0.2.30).

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:25]: Set it to unknown.

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:12]: Please update to the latest (https://github.com/argoproj-labs/argocd-autopilot/releases/tag/v0.2.30).

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:15]: Please update to the latest (https://github.com/argoproj-labs/argocd-autopilot/releases/tag/v0.2.30).

**Review by bryanasdev000 — CHANGES_REQUESTED** [review]
LGTM, just update it and remove version.

Thanks for the awesome work!

NOTE: You probably will need to a rebase too, just in case.

**Comment by bryanasdev000** [comment]
For reference https://github.com/NixOS/nixpkgs/pull/163522 .

**Comment by sagikazarmark** [comment]
@bryanasdev000 Thanks for the review. I made the requested changes.

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:14]: Hmm I think the correct sha is `sha256-0bWmmZ0eGLHRzkNU75qg7GS3XfsR7p84IBNluEtk8Es=` (local built). Other than that, everything seems OK.

- https://github.com/NixOS/nixpkgs/pull/147388/checks?check_run_id=5496113280

**Review by bryanasdev000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (sagikazarmark) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix:14]: That doesn't work for me on macOS:

Képernyőfotó 2022-03-11 - 10 32 28

Comment (bryanasdev000) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix:14]: I have no idea about macOS, but maybe is something along these lines?

https://github.com/golang/go/issues/29282

Could you clean nix cache (`nix-collect-garbage -d; nix-store --optimise;`) and enable sandbox for nix (se OP)?

**Comment by bryanasdev000** [comment]
@sagikazarmark, a new release of argocd-autopilot is out, can you please test these points?

- update to 0.3.0, I got this shas on a Nix single user install in Ubuntu:

```
 sha256 = "sha256-tggE1T+oD/dJS9tD9xOExjhy+T1GDd0vwTerD3P2KvA=";
 vendorSha256 = "sha256-UqZB3PknZrcgClZjpuFr9iLu1netwauTg/LpWoIVB0I="; 
```

- add back commit + tag, so we can have a little standard between argo expressions (you can check https://github.com/NixOS/nixpkgs/pull/163522, once we create a updateScript we replicate it for both expressions)

- add myself as maintainer too

**Comment by sagikazarmark** [comment]
@bryanasdev000 made the requested changes

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (sagikazarmark) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix]: Did all these, enabled sandbox and I'm still getting a hash mismatch for the modules.

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:12]:
```suggestion

  src = fetchFromGitHub {
    owner = "argoproj-labs";
    repo = "argocd-autopilot";
    rev = "v${version}";;
```

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:24]:
```suggestion
      "-X ${package_url}.version=${src.rev}"
```

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:28]:
```suggestion
      "-X ${package_url}.installationManifestURL=github.com/argoproj-labs/argocd-autopilot/manifests/base?ref=${src.rev}"
      "-X ${package_url}.installationManifestsNamespacedURL=github.com/argoproj-labs/argocd-autopilot/manifests/insecure?ref=${src.rev}"
```

**Review by bryanasdev000 — APPROVED** [review]
> @bryanasdev000 made the requested changes

Thanks for the awesome work!

I was going to ask about SHA on macOS but by the tests I've seen that unfortunately they didn't pass :(

That's fine by me, you'll probably have to make the adjustments that @SuperSandro2000 asked for (he will provide much better feedback than mine \o/)

I'd like to resolve the Mac issues, but from what I've researched and talked about, most likely current macOS support (Tier 3) and update issues will likely block this PR if we wait.

Some links about:

- https://github.com/NixOS/nixpkgs/pull/127519
- https://github.com/NixOS/nixpkgs/pull/164320
- https://discourse.nixos.org/t/missing-go-1-17-on-x86-64-darwin/15032/3

NOTE: We may need to use `platforms` to limit the use of this package to Linux only, example: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/jmeter/default.nix#L58 .

**Comment by bryanasdev000** [comment]
> @bryanasdev000 made the requested changes

Thanks for the awesome work!

I was going to ask about SHA on macOS but by the tests I've seen that unfortunately they didn't pass :(

That's fine by me, you'll probably have to make the adjustments that Super asked for (it will provide much better feedback than mine \o/)

I'd like to resolve the Mac issues, but from what I've researched and talked about, most likely current macOS support (Tier 3) and update issues will likely block this PR if we wait.

Some links about:

- https://github.com/NixOS/nixpkgs/pull/127519
- https://github.com/NixOS/nixpkgs/pull/164320
- https://discourse.nixos.org/t/missing-go-1-17-on-x86-64-darwin/15032/3

NOTE: We may need to use `platforms` to limit the use of this package to Linux only, example https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/jmeter/default.nix#L58 .

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (sagikazarmark) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix:12]: This removes the commit and adds an extra semi-colon at the end of rev.

I'm going to make these changes manually, but @bryanasdev000 asked me to add back the commit. So what should it be?

**Comment by sagikazarmark** [comment]
Well, that didn't solve the problem on Mac, but it was necessary on my NixOS to build properly.

**Comment by bryanasdev000** [comment]
> Well, that didn't solve the problem on Mac, but it was necessary on my NixOS to build properly.

Strange, it build with success in my NixOS host running unstable.

Which branch are you running?

Also, based in the ofborg output we will really need to restrict platforms... :(

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix]: > This removes the commit and adds an extra semi-colon at the end of rev.
> 
> I'm going to make these changes manually, but @bryanasdev000 asked me to add back the commit. So what should it be?

I think by default, a lot of packages Go goes without it, but based on @thiagokokada feedback on https://github.com/NixOS/nixpkgs/pull/163522 we need it to make sure that the application does not behave out of ordinary.

Also, to bot update it will need to work on a update script or else we will need to update it by hand each time.

**Comment by sagikazarmark** [comment]
> Which branch are you running?

21.11

> Also, based in the ofborg output we will really need to restrict platforms... :(

Can't we just use different checksums for different platforms?

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (sagikazarmark) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix]: I'm fine with either, just tell me which one should I leave in this PR (we can always add back the commit later though).

**Comment by bryanasdev000** [comment]
> > Which branch are you running?
> 
> 21.11
> 
> > Also, based in the ofborg output we will really need to restrict platforms... :(
> 
> Can't we just use different checksums for different platforms?

Indeed! The bot probably won't auto update it but it may works, checkout this snippets.

For Go1.17 (file `pkgs/top-level/all-packages.nix`)

```
  argocd-autopilot = callPackage ../applications/networking/cluster/argocd-autopilot {
       buildGoModule = buildGo117Module;
  };
```

NOTE: Leave the header as buildGoModule instead of buildGo117Module.

For the vendor (file pkgs/applications/networking/cluster/argocd-autopilot/default.nix`):

```
{ lib, stdenv, buildGoModule, fetchFromGitHub }:
...
vendorSha256 = if stdenv.hostPlatform.system == "x86_64-darwin" then "" else "sha256-UqZB3PknZrcgClZjpuFr9iLu1netwauTg/LpWoIVB0I=";
...
```

Can you try and report?

Thanks!

**Comment by bryanasdev000** [comment]
EDIT: I successfully built it on an Ubuntu host with Nix single user with above changes to your latest changes.

The "bad" thing is who going to update it need to provide one SHA, make the PR and get the SHA of the other platform based on ofborg (or you're going to have both platforms with you :P)

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix]: > I'm fine with either, just tell me which one should I leave in this PR (we can always add back the commit later though).

I am with you, but the call is basically from who merges it.

In the last update of ArgoCD I dropped it, but on @thiagokokada  comments I added it.

In the end, a updateScript will be needed for this package as long as we have different hashes, so may vote for now will probably rest on maintain this with commit and maintain some "default" for argo-* expressions.

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (jonringer) [inline pkgs/applications/networking/cluster/argocd-autopilot/default.nix:15]: for some reason, darwin doesn't like this. Generally this is from unicode being in file names or two file paths only differing by capitalizaiton.

```
installing
hash mismatch in fixed-output derivation '/nix/store/sfc82855k4qxig2lk20jrvfmmsc6mw6w-argocd-autopilot-0.3.0-go-modules':
  wanted: sha256:0hh72n15msgjhf9sphddfzbfw8pndghscqsn18hbfri7z7f439jj
  got:    sha256:13020nyz9jp3hghaziz9sqnvbk6spwm5xmxbhdjyv1sspj4xawsi
cannot build derivation '/nix/store/hxc62vf7k5l007pfs7y77g9wdyqfg114-argocd-autopilot-0.3.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/hxc62vf7k5l007pfs7y77g9wdyqfg114-argocd-autopilot-0.3.0.drv' failed
```

**Comment by sagikazarmark** [comment]
Unfortunately an empty vendor sha doesn't work.

Let me try to find the offending library, maybe we can fix it upstream.

**Comment by SuperSandro2000** [comment]
> Unfortunately an empty vendor sha doesn't work.
> 
> Let me try to find the offending library, maybe we can fix it upstream.

I would highly prefer that over the if else. Maybe proxyVendor already helps?

File: pkgs/applications/networking/cluster/argocd-autopilot/default.nix
Comment (bryanasdev000) [reply pkgs/applications/networking/cluster/argocd-autopilot/default.nix]: > for some reason, darwin doesn't like this. Generally this is from unicode being in file names or two file paths only differing by capitalizaiton.
> 
> ```
> installing
> hash mismatch in fixed-output derivation '/nix/store/sfc82855k4qxig2lk20jrvfmmsc6mw6w-argocd-autopilot-0.3.0-go-modules':
>   wanted: sha256:0hh72n15msgjhf9sphddfzbfw8pndghscqsn18hbfri7z7f439jj
>   got:    sha256:13020nyz9jp3hghaziz9sqnvbk6spwm5xmxbhdjyv1sspj4xawsi
> cannot build derivation '/nix/store/hxc62vf7k5l007pfs7y77g9wdyqfg114-argocd-autopilot-0.3.0.drv': 1 dependencies couldn't be built
> error: build of '/nix/store/hxc62vf7k5l007pfs7y77g9wdyqfg114-argocd-autopilot-0.3.0.drv' failed
> ```

Any tip to track down the case?

Also TIL for MacOS...

**Comment by bryanasdev000** [comment]
> Unfortunately an empty vendor sha doesn't work.
> 
> Let me try to find the offending library, maybe we can fix it upstream.

Oh the empty SHA was only because I did not find the MacOS one.

**Comment by bryanasdev000** [comment]
> > Unfortunately an empty vendor sha doesn't work.
> > Let me try to find the offending library, maybe we can fix it upstream.
> 
> I would highly prefer that over the if else. Maybe proxyVendor already helps?

Reading the expression for Go this maybe the case...

But how it works? We put a SHA that's working for Linux and it does magic?

**Comment by SuperSandro2000** [comment]
go proxy is a feature of go which does not download the dependencies via git but a proxy on the web which does some magic.

**Comment by bryanasdev000** [comment]
> go proxy is a feature of go which does not download the dependencies via git but a proxy on the web which does some magic.

I created a PR just to use ofborg to test this, let's see!

**Comment by bryanasdev000** [comment]
It worked! All checks green! Thanks for the tip @SuperSandro2000 !

- https://github.com/NixOS/nixpkgs/pull/166015/checks 
- https://github.com/NixOS/nixpkgs/pull/166015/checks?check_run_id=5712933241
- https://github.com/NixOS/nixpkgs/pull/166015/checks?check_run_id=5712916405
- https://github.com/NixOS/nixpkgs/pull/166015/checks?check_run_id=5712955567
- https://github.com/NixOS/nixpkgs/pull/166015/checks?check_run_id=5712917294
- https://logs.nix.ci/?key=nixos/nixpkgs.166015&attempt_id=0f2a53b7-f1ad-4f3c-9ee1-15d98f5fb642
- https://logs.nix.ci/?key=nixos/nixpkgs.166015&attempt_id=be23d6c0-8934-4b84-9330-c904c332d8cc
- https://logs.nix.ci/?key=nixos/nixpkgs.166015&attempt_id=e5449d1a-5bc5-4f70-ab5d-c05fa5fd4149
- https://logs.nix.ci/?key=nixos/nixpkgs.166015&attempt_id=3cd2c1e5-797a-41f7-a39e-2512b37f48d2

@sagikazarmark could you please check and add this snipets?

```
vendorSha256 = "sha256-v8UMSObE6w+ULzueEK0UFeebLqoamy/788SQLBmJZ8U=";

proxyVendor = true;
```

Full expression in https://github.com/NixOS/nixpkgs/pull/166015/commits/187cbbf80eafae139a6c5d8ce056943acd73b893

**Comment by sagikazarmark** [comment]
Updated the PR. Thanks @bryanasdev000 !

File: pkgs/top-level/all-packages.nix
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:24561]: Isn't this the default already 🤔 ?

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:24561]: Good catch! It is by now. Can be safely removed.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 147388` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>argocd-autopilot</li>
  </ul>
</details>

File: pkgs/top-level/all-packages.nix
Comment (bryanasdev000) [reply pkgs/top-level/all-packages.nix:24561]: I think we left it as is because of @sagikazarmark's comment, and points like:

- Go 1.17 is currently used by the application;
- Go 1.18 is being implemented in nixpkgs;
- No idea about the default for darwin;
- Maybe backport to 21.11 need this;

**Comment by bryanasdev000** [comment]
Happy to see this merged!

Thanks @sagikazarmark  @SuperSandro2000  @thiagokokada  @jonringer!

File: pkgs/top-level/all-packages.nix
Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:24561]: AFAIK:

- We don't necessary use the same Go version as upstream unless there is a reason for it (e.g.: doesn't build with a newer version)
- The default for darwin should be the same
- On 21.11 backport you can set this

**Pull request merged**
